sklearn/preprocessing/data.py

```diff
@@ -1811,138 +1813,248 @@ class OneHotEncoder(BaseEstimator, TransformerMixin):
-                                dtype=self.dtype).tocsr()
+        self.one_hot_feature_index_ = np.empty(output_cols, dtype=np.int)
 
-        if (isinstance(self.n_values, six.string_types) and
-                self.n_values == 'auto'):
-            mask = np.array(out.sum(axis=0)).ravel() != 0
-            active_features = np.where(mask)[0]
-            out = out[:, active_features]
-            self.active_features_ = active_features
+        for i in range(n_features):
+            s, e = self.feature_index_range_[i]
+            self.one_hot_feature_index_[s:e] = i
 
-        return out if self.sparse else out.toarray()
+        return self
 
-    def fit_transform(self, X, y=None):
-        """Fit OneHotEncoder to X, then transform X.
+    def _fit(self, X):
+        "Assumes `X` contains only catergorical features."
 
-        Equivalent to self.fit(X).transform(X), but more convenient and more
-        efficient. See fit for the parameters, transform for the return value.
+        if not np.issubdtype(X.dtype.type, np.integer):
+            X = check_array(X, dtype=np.object)
+        n_samples, n_features = X.shape
+
+        self._n_features = n_features
+        self._label_encoders = [LabelEncoder() for i in range(n_features)]
+        # Maximum value for each featue
+        self._max_values = [None for i in range(n_features)]
+
+        if self.n_values is not None:
+            warnings.warn('The parameter `n_values` is deprecated, use the'
+                          'parameter `values` instead and specify the '
+                          'expected values for each feature')
+
+            if isinstance(self.n_values, numbers.Integral):
+                if (np.max(X, axis=0) >= self.n_values).any():
+                    raise ValueError("Feature out of bounds for n_values=%d"
+                                     % self.n_values)
+                self.values = self.n_values
+            else:
+                try:
+                    n_values = np.asarray(self.n_values, dtype=int)
+                except (ValueError, TypeError):
+                    raise TypeError("Wrong type for parameter `n_values`."
+                                    " Expected 'auto', int or array of ints,"
+                                    "got %r" % type(X))
+                if n_values.ndim < 1 or n_values.shape[0] != X.shape[1]:
+                    raise ValueError("Shape mismatch: if n_values is an array,"
+                                     " it has to be of shape (n_features,).")
+                self.values = list(self.n_values)
+
+        error_msg = ("`values` should be 'auto', an integer, a list of"
+                     " integers or a list of list")
+
+        for i in range(n_features):
+            le = self._label_encoders[i]
+
+            self._max_values[i] = np.max(X[:, i])
+            if self.values == 'auto':
+                le.fit(X[:, i])
+            elif isinstance(self.values, numbers.Integral):
```

Comment:
use `np.isscalar`? There are problems using `numbers.Integral` for numpy int types, iirc.